syntax = "proto3";
package CarCommunicationApp;

// ======== BOTH ===============

message CarParameters {
    float steer = 1;
    float forward = 2;
    float backward = 3;
}

// ======== SERVER ===============

message ServerResponse {
    enum Command {
        LEFT = 0;
        RIGHT = 1;
        FORWARD = 2;
        BACKWARD = 3;
        STOP = 4;
        NOOP = 5; 
        POWEROFF = 6;
        RESPAWN = 7;
        MOVEMENT = 8;
    }
    
    Command command = 1;
    CarParameters car_parameters = 2;
}

// ======== CLIENT ===============

message DistanceSensorsData {
  float front_left_distance = 1;
  float front_distance = 2;
  float front_right_distance = 3;
  float back_left_distance = 4;
  float back_distance = 5;
  float back_right_distance = 6;
}

message RouterData {
    string id = 1;
    float rssi = 2;
}

message ClientRequest {
    // real model data
    string car_id = 1;
    float car_speed = 2;
    CarParameters car_parameters = 3;
    bytes camera_image = 4;
    DistanceSensorsData distance_sensors_data = 5;
    repeated RouterData routers_data = 6; 
    // virtual model data
    bool boxes_in_camera_view = 7; 
    bool car_collision_data = 8;
    string qr_code_metadata = 9;
}

service Communication {
  rpc SendRequest(ClientRequest) returns (ServerResponse);
}
